stripe: Add cron-based plan invoicing to remote realm/server billing system.

prakhar1144 opened the pull request:
Email to Zulip team:
![image](https://github.com/zulip/zulip/assets/56781761/8779409d-8b1b-4efc-a5a7-81e6d3143a2e)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented on corporate/lib/stripe.py:
```diff
@@ -4255,14 +4272,59 @@ def get_plan_renewal_or_end_date(plan: CustomerPlan, event_time: datetime) -> da
+            #     to_emails=[remote_server.contact_email],
+            #     from_address=BILLING_SUPPORT_EMAIL,
+            #     context=context,
+            # )
```
Let's hold off on emailing the admin for now. We'll just have a manual process for contacting the admin; we can automate it if/when we know better what the causes and remedies are.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -2394,6 +2394,23 @@ def invoice_plan(self, plan: CustomerPlan, event_time: datetime) -> None:
+                    description=f"${flat_discount}/month new customer discount",
+                    # Negative value to apply discount.
+                    amount=(-1 * discount),
+                )
```
Do we have `@transaction.atomic` around `invoice_plan`? reading this code, it would definitely be very nice if we could be certain that the parts of this happen transactionally, and I imagine that will be true for other parts of invoicing as well.

prakhar1144 replied:
Second commit.

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -4255,14 +4272,59 @@ def get_plan_renewal_or_end_date(plan: CustomerPlan, event_time: datetime) -> da
+    assert server.last_audit_log_update is not None
+
+    # 'last_audit_log_update' is updated hourly.
+    return event_time - server.last_audit_log_update >= timedelta(days=1)
```
I think the right algorithm is that:
* We should never invoice a plan where `server.last_audit_log_update < next_invoice_date` -- but we don't need to throw an exception if that temporal difference is small.
* If that temporal difference is greater than a day, we should send the email to our support team to investigate.
* We probably need some way to track the last date for which we sent such an email -- we don't want to generate such an email every hour/day.

prakhar1144 replied:
Oh, I think there was a miscommunication. The issue states:

> Probably we want to complain in a way that the Zulip team will see it whenever we are due for invoicing and that is overdue by more than a day.

**whenever we are due for invoicing and that is overdue by more than a day.**

I thought it as "one day has passed since the date (next_invoice_date) invoicing should occur"

prakhar1144 replied:
> We probably need some way to track the last date for which we sent such an email

Added the following field in CustomerPlan:
```python
    # Flag to track if an email has been sent to Zulip team for
    # invoice overdue by >= one day. Helps to send an email only once
    # and not every time when cron run.
    invoice_overdue_email_sent = models.BooleanField(default=False)
```
Also, verified the behaviour in tests.

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comments -- see respective comments
* Added a commit to fix a small variable renaming issue (first commit)

prakhar1144 commented:
Updates in the recent push:
* Addressed the review comments -- see respective comments
* Added a commit to fix a small variable renaming issue (first commit)

timabbott commented on corporate/lib/stripe.py:
```diff
@@ -2347,6 +2347,7 @@ def switch_plan_tier(self, current_plan: CustomerPlan, new_plan_tier: int) -> No
             licenses_at_next_renewal=licenses_for_new_plan,
         )
 
+    @transaction.atomic
```
Hmm, this change might be dangerous. The problem is that the Stripe operations are side effects that don't happen inside the Zulip database, and so having this whole thing be a transaction means that it's possible for changes to happen in Stripe without appearing in Zulip.

I don't really know what the best answer for this is -- but usually it involves multiple commit phases.

timabbott commented:
This looks good without the transaction commit; posted in Zulip but I think probably that commit should just be its own future project.

Merged, thanks for all the work on this @prakhar1144!
